### PR TITLE
Support -trait and -notrait for RunTest.exe

### LINKED ISSUE #6241: Support -trait for RunTest.exe
[Roslyn.Services.Editor.(\w+).UnitTests are known slow](https://github.com/dotnet/roslyn/commit/c3e4799eb2a5fb5fb84ff9a938868bfd177b8fd3#diff-0102553256b475e6f1e0f2663dedf27bL93), it would be nice that `-trait` can be specified to shorten the test time when running them locally, as running test is the easy way for external contributors to verify editor features currently, or I have to use the long xunit runner command lines and if failure encountered, have to open the browser myself after finding the report file.

### PR DESCRIPTION
**Pull request by oocx**
RunTests has additional arguments to run tests
with specific traits. These arguments can also be
specified as build properties for BuildAndTest.proj.
Example: msbuild BuildAndTest.proj /p:Trait="Feature=Classification". 
I also refactored parsing of the command line arguments into it's own class.

Fix #6241

### CODE REVIEW

File: src/Tools/Source/RunTests/Options.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+            if (opt.Assemblies.Length != 0) return opt;
+
+            PrintUsage();
+            return null;
+        }        
+
+        private static void PrintUsage()
+        {
+            Console.WriteLine("runtests [xunit-console-runner] [-test64] [-xml] [-trait:listOfTraitsToInclude] [-notrait:listOfTraitsToExclude] [assembly1] [assembly2] [...]");
```
Comment (diryboy) [inline src/Tools/Source/RunTests/Options.cs]: Would be nice if example/separator included. Like `[-trait:name1=value1;..]`.

File: src/Tools/Source/RunTests/Options.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                    index++;
+                }
+                else if (comp.Equals(current, "-xml"))
+                {
+                    opt.UseHtml = false;
+                    index++;
+                }
+                else if (current.StartsWith("-trait:", StringComparison.OrdinalIgnoreCase))
+                {
+                    opt.Trait = current.Substring(7);
```
Comment (diryboy) [inline src/Tools/Source/RunTests/Options.cs]: If the argument is just `-trait:`, this crash, similar below.

**Comment by oocx** [comment]
Thank you for your feedback.

I have added additional error handling for the command line arguments. Only handling the "trait:" case would only fix the problem in Options.cs, but still cause an exception later in the program as this would add "trait:" to the assemblies list and try to read it as a file.

I also changed the PrintUsage output so that it should be easier to see how multiple traits can be specified.

**Comment by davkean** [comment]
Added the failure for unit64 to https://github.com/dotnet/roslyn/issues/6166.

retest prtest/win/dbg/unit64

**Comment by davkean** [comment]
test eta please
retest this please

**Comment by davkean** [comment]
tag @dotnet/roslyn-infrastructure  for review

**Comment by oocx** [comment]
I've been thinking about this some more. With the additional validation I added in the last commit, the tests would not run at all if one of the specified test assemblies is missing. That's different than the old behavior. Should I leave it like that or remove the File.Exists? If I remove it, I would probably have to replace it with a check that verifies that the file name is at least valid, otherwise we would be back to the previous problem that invalid command line arguments like "-trait:" (without anything after the ':') are added to the assembly list.

**Comment by davkean** [comment]
Adding @dotnet/roslyn-ide also to take a look.

**Comment by dpoeschl** [comment]
Previous prtest/lin/dbg/unit32 test run failed (http://dotnet-ci.cloudapp.net/job/roslyn_prtest_lin_dbg_unit32/552/):

```
10:57:57 Native stacktrace:
10:57:57 
10:57:57 mono_class_from_mono_type: implement me 0x9d
10:57:57 
10:57:57 * Assertion: should not be reached at class.c:6523
10:57:57 
10:57:57 
./cibuild.sh: line 234: 17794 Aborted                 (core dumped) mono $MONO_ARGS $xunit_runner Binaries/$BUILD_CONFIGURATION/$i.dll -xml Binaries/$BUILD_CONFIGURATION/xUnitResults/$i.dll.xml -noshadow
```

Filed https://github.com/dotnet/roslyn/issues/6348 to track that. Let's see if it fails that way again...

retest prtest/lin/dbg/unit32 please

**Comment by dpoeschl** [comment]
retest prtest/lin/dbg/unit32 please

**Comment by dpoeschl** [comment]
retest prtest/lin/dbg/unit32 please

File: src/Tools/Source/RunTests/Options.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                {
+                    opt.Test64 = true;
+                    index++;
+                }
+                else if (comp.Equals(current, "-xml"))
+                {
+                    opt.UseHtml = false;
+                    index++;
+                }
+                else if (current.Length > 7 && current.StartsWith("-trait:", StringComparison.OrdinalIgnoreCase))
```
Comment (Pilchie) [inline src/Tools/Source/RunTests/Options.cs]: You could re-use `comp` here (or inline it everywhere).

**Comment by Pilchie** [comment]
Tagging @tannergooding @agocke @jaredpar explicitly.  What do you think about @oocx 's question about the `File.Exists` check?

**Comment by ManishJayaswal** [comment]
@oocx I think it makes sense to run the rest of the tests even if one or more assemblies are missing. At the end we should report the missing assemblies.

File: src/Tools/Source/RunTests/Options.cs (hunk 3 above)
Comment (oocx) [inline src/Tools/Source/RunTests/Options.cs]: comp is a StringCompar**er**, StartsWith expects a StringCompar**ison**. I don't see how I could reuse comp with StartsWith.

**Comment by oocx** [comment]
RunTests will now run even if some of the assemblies are missing / invalid. If assemblies are missing / invalid, it will output a list of the missing assemblies after running the tests.

File: src/Tools/Source/RunTests/Options.cs (hunk 3 above)
Comment (Pilchie) [inline src/Tools/Source/RunTests/Options.cs]: Ah yes.  Missed that, sorry.

Comment (oocx) [inline src/Tools/Source/RunTests/Options.cs]: No problem :).

**Comment by davkean** [comment]
Thanks - sorry this got missed, can you resolve conflicts and resubmit - and we'll get this through.

**Comment by oocx** [comment]
Done. I hope I merged correctly, I have never done that with git before.

**Comment by davkean** [comment]
Merging this in, thanks for your contribution!

**Comment by diryboy** [comment]
thanks!

**Pull request merged**